src/main/java/eu/arrowhead/application/skeleton/consumer/ConsumerMain.java

```diff
@@ -0,0 +1,435 @@
+		} catch (final Exception e) {
+			logger.error(e.getMessage());
+		}
+
+		for (final Long id : ruleIdsToDelete) {
+			logger.debug("Removing authorization rule with id: " + id);
+			final String response = arrowheadService.consumeServiceHTTP(String.class, HttpMethod.DELETE,
+					Utilities.createURI(authorizationUri.get(ConsumerConstants.SCHEME), authorizationUri.get(ConsumerConstants.HOST),
+							Integer.parseInt(authorizationUri.get(ConsumerConstants.PORT)), authorizationUri.get(ConsumerConstants.PATH)
+									+ ConsumerConstants.QUERY_AUTH_INTRA_CLOUD + "/" + Long.toString(id)),
+					null, null);
+			logger.debug("Http DELETE response: " + response);
+		}
+	}
+
+	// -------------------------------------------------------------------------------------------------
+	private void addRules(final List<AuthRule> rules) {
+
+		final List<AuthorizationIntraCloudRequestDTO> rulesToAdd = createDTOListFromAuthRules(rules);
+		for (final AuthorizationIntraCloudRequestDTO ruleToAdd : rulesToAdd) {
+			addSingleRule(ruleToAdd);
+		}
+	}
+
+	// -------------------------------------------------------------------------------------------------
+	private List<AuthorizationIntraCloudRequestDTO> createDTOListFromAuthRules(final List<AuthRule> rules) {
+
+		final List<AuthorizationIntraCloudRequestDTO> dtoList = new ArrayList<>();
+
+		List<Long> consumerIdList; //multiple consumerid can match one auth rule's systeminfo
+		List<Long> prodiverIdList; //multiple providerid can match one auth rule's systeminfo
+
+		List<Long> serviceDefinitionId; //contains one element, but must be a list because of the DTO
+		List<Long> interfaceIds;
+
+		for (int i = 0; i < rules.size(); i++) {
+
+			consumerIdList = new ArrayList<>();
+			prodiverIdList = new ArrayList<>();
+			serviceDefinitionId = new ArrayList<>(1);
+			interfaceIds = new ArrayList<>();
+
+			try {
+				consumerIdList = getSystemIdsByInfo(rules.get(i).getConsumer());
+				prodiverIdList = getSystemIdsByInfo(rules.get(i).getProvider());
+
+				serviceDefinitionId.add(serviceDefinitionToId(rules.get(i).getService()));
+
+				for (final String interfaceName : rules.get(i).getInterfaces()) {
+					interfaceIds.add(interfaceNameToId(interfaceName));
+
+				}
+
+				for (final Long consumerId : consumerIdList) {
+					for (final Long providerId : prodiverIdList) {
+						dtoList.add(new AuthorizationIntraCloudRequestDTO(consumerId, Arrays.asList(providerId), serviceDefinitionId,
+								interfaceIds));
+					}
+				}
+			} catch (final Exception e) {
+				logger.error("Could not create authorization rule: " + rules.get(i).toString() + ", reason: "
+						+ e.getMessage());
+			}
+		}
+		return dtoList;
+
+	}
+
+	// -------------------------------------------------------------------------------------------------
+	private void addSingleRule(final AuthorizationIntraCloudRequestDTO ruleToAdd) {
+		Map<String, String> authorizationUri = null;
+		try {
+			authorizationUri = getAuthorizationUri();
```

Comment:
Authorization URI stays the same, once you have retrieved it. No need for always go through on all the systems to find it. Should be cached  

Plus, if you haven't found the uri, you shoud terminate.